lib/iris/cube.py

```diff
@@ -1681,43 +1699,52 @@ def data(self):
             (10, 20)
 
         """
-        data = self._my_data
-        if is_lazy_data(data):
+        if self._numpy_array is None:
             try:
-                data = as_concrete_data(data)
+                data = self._dask_array.compute()
+                mask = np.isnan(data)
+                if data.dtype != self.dtype:
+                    data = data.astype(self.dtype)
+                    self.dtype = None
+                if np.all(~mask):
+                    self._numpy_array = data
+                else:
+                    fv = self.fill_value
+                    self._numpy_array = ma.masked_array(data, mask=mask,
+                                                        fill_value=fv)
             except MemoryError:
                 msg = "Failed to create the cube's data as there was not" \
                       " enough memory available.\n" \
                       "The array shape would have been {0!r} and the data" \
                       " type {1}.\n" \
                       "Consider freeing up variables or indexing the cube" \
                       " before getting its data."
-                msg = msg.format(self.shape, data.dtype)
+                msg = msg.format(self.shape, self.dtype)
                 raise MemoryError(msg)
-            # Unmask the array only if it is filled.
-            if (isinstance(data, np.ma.masked_array) and
-                    ma.count_masked(data) == 0):
-                data = data.data
-            # data may be a numeric type, so ensure an np.ndarray is returned
-            self._my_data = np.asanyarray(data)
-        return self._my_data
+        return self._numpy_array
 
     @data.setter
     def data(self, value):
-        data = np.asanyarray(value)
+        if not (hasattr(value, 'shape') and hasattr(value, 'dtype')):
+            value = np.asanyarray(value)
 
-        if self.shape != data.shape:
+        if self.shape is not None and self.shape != value.shape:
             # The _ONLY_ data reshape permitted is converting a 0-dimensional
             # array i.e. self.shape == () into a 1-dimensional array of length
             # one i.e. data.shape == (1,)
-            if self.shape or data.shape != (1,):
+            if self.shape or value.shape != (1,):
                 raise ValueError('Require cube data with shape %r, got '
-                                 '%r.' % (self.shape, data.shape))
+                                 '%r.' % (self.shape, value.shape))
 
-        self._my_data = data
+        if is_lazy_data(value):
+            self._dask_array = value
+            self._numpy_array = None
+
+        else:
+            self._numpy_array = value
 
     def has_lazy_data(self):
-        return is_lazy_data(self._my_data)
+        return True if self._numpy_array is None else False
```

Comment:
> Also, isn't it the case the both cube._numpy_array and cube._dask_array could be populated? 
yes, at present this is the case
> If so, are we saying that if that's the case, the cube isn't lazy?
yes, it is not lazy
if `cube._numpy_array` exists it is always used in preference

> Is there an explicit statement to define whether it's valid for both of these attributes to be populated? 
No, there is no explicit statement.  It is an open question.  
We should answer it.  The question here is should we answer it on this PR or raise an issue and answer it?
The utility of having both is the `cube._numpy_array` may be set to None, reverting the cube back to it's initally loaded state, with a file proxy.  I don't know how valuable this may be.

> Sutely, it's totally invalid for both not to be populated ... 

Yes, that is invalid and exceptions are thrown

> do we have checks and balances for that to maintain that integrity within the cube?

there are some, are there enough?